### PR TITLE
Add image files ending with star (*) instead of specifying the format

### PR DESCRIPTION
**Pull request by jjcaballero**
**Some context:**
I was trying to include plots generated by matplotlib.sphinext when using the confluence builder. Then I learned that plots were properly generated and the only problem was the matplotlib.sphinext extension was not considering confluence as a possible builder.
Therefore, I created this PR (https://github.com/matplotlib/matplotlib/pull/14681) there, to ask them to include it. Eventually, this was addressed in other PR (https://github.com/matplotlib/matplotlib/pull/14683) which enhance the same but in a more general way. Now, builders which want to include plots from matplotlibs should handle images ending with star. 

And handling image with star is the goal of this PR. I have added the parameter **supported_image_types** to the builder like here:  [sphinx builders doc](https://www.sphinx-doc.org/en/master/usage/builders/index.html#sphinx.builders.html.StandaloneHTMLBuilder.supported_image_types) 
We might want to include more image formats, I have just considered png, jpeg and pdf.

**In a nutshell:** 
when the image ends with star and does not exists, then the ConfluenceAssetManager will use the first found image with the same name but ending with a supported image type.

**Side note:** maybe I am wrong, but I think there are no tests to verify the publishing itself.  Is there any recommendation to do such a thing? It would be very convenient to be able to run tests choosing confluence server and parent page. I mean using the same infrastructure used to run tests but publishing.

### CODE REVIEW

**Comment by jdknight** [comment]
To start off, it is most likely that #228 will be merged in before this change (which will cause conflicts; but I do not mind correcting these conflicts if needed).

I have no problem with the idea of supporting a wildcard to map to detected assets which may be generated by other extension. I would like to also try this out to actually see it being performed (even though the idea makes sense after reading the referenced PRs). I am a bit hesistent in having this `supported_image_types` list though. Sphinx does have some mime definitions inside their `images.py` implementation -- I wonder if something would be doing with this to prevent this extension to trying to manage supported image types (although, I have no problem definition a separate independent list to whitelist or blacklist image types based off official documented Confluence supported image types.

Note that due to other issues being addressed in this extension,  the expected window for this feature is planned around end-of-year (2019; see [this comment](https://github.com/sphinx-contrib/confluencebuilder/pull/215#issuecomment-541268554) for more details).

**Comment by jessetan** [comment]
I have a similar usecase. I build both HTML and PDF output from a single source. To make images (especially diagrams) display as crisp as possible, I use SVG diagrams for HTML output and PDF diagrams for inclusion into PDF output.
The image is included using `.. figure:: images/diagram.*`. The actual file (`diagram.svg` or `diagram.pdf`) that is included depends on the builder that is used.

Sphinx already has code that handles an asterisk in the file extension: https://github.com/sphinx-doc/sphinx/blob/01f8dad160d206a48302b1bcd91b4e0bc36e8a10/sphinx/builders/__init__.py#L186. This replaces the `*` in the filename based on the `supported_image_types` of a writer.
This is called by the default writers in Sphinx but not by confluencebuilder.I can make my usecase work by adding a suitable `supported_image_types` and adding a call to `self.post_process_images(doctree)` in `prepare_writing()` here: https://github.com/sphinx-contrib/confluencebuilder/blob/01def8c69a51234c5392a69b0022492e950531ed/sphinxcontrib/confluencebuilder/builder.py#L236
(The exact line number is not important, but it needs to run on every documents doctree before the call `self.assets.process(ordered_docnames)`.

@jjcaballero would that solve your usecase? It is much less code than this PR and will keep maintenance of the code that finds the correct file the responsibility of Sphinx, not confluencebuilder.

> I wonder if there is something we could be doing with this to prevent this extension to trying to manage supported image types

@jdknight As Confluence content ends up as HTML anyway, one could assume that the image formats that are supported are as the same for the HTML builder. The list of supported images can then be defined in code with:
```py
from sphinx.builders.html import StandaloneHTMLBuilder 

...

class ConfluenceBuilder(Builder):
    supported_image_types = StandaloneHTMLBuilder.supported_image_types
```

**Comment by jjcaballero** [comment]
@jessetan your approach seems cleaner and works for my use case too.
Do you plan to do a PR for that or should I include it here? 
I would like to try to keep the expected window for this feature mentioned by @jdknight (end of the year 2019)

**Comment by jessetan** [comment]
@jjcaballero if you are willing, it would be great if you could make the changes in this PR. It would save me some work and you have the testcases ready.

After my previous comment, I've found the best place to call `self.env.app.builder.post_process_images(doctree)` is before `processDocument` in https://github.com/sphinx-contrib/confluencebuilder/blob/046ac21f7528b5777826f1c1d21a63bd347f091c/sphinxcontrib/confluencebuilder/assets.py#L138

So far I have found only one downside; builds fail if the PlantUML Sphinx extension is used (https://github.com/sphinx-contrib/plantuml/). The image nodes inserted by that plugin never get a `candidates` attribute.
I've worked around this by adding the following before the call to `post_process_images`
```py
        # Verify every image has a 'candidates' attribute with key '*'
        # This is normally added by Sphinx' ImageCollector, but is not called
        # by sphinxcontrib.plantuml for some reason
        # https://github.com/sphinx-doc/sphinx/blob/089046979f2b84cf850ee95776d10fe9fd9a77e9/sphinx/environment/collectors/asset.py#L43
        for node in doctree.traverse(imageNode):
            if 'candidates' not in node:
                node['candidates'] = { '*': node['uri'] }
```

There must be a better way to do this, but I haven't had time to delve into this issue any deeper. It's probably not necessary to try to fix that in this PR.

**Comment by jjcaballero** [comment]
@jdknight the pull request should be ready now to review/merge. Thanks.

@jessetan eventually, I placed  `self.post_process_images(doctree)`  in  ``prepare_writing() ``  from builder.py. As this method belongs to the parent class sphinx.Builder, IMO it is clearer to call it directly from the child class ConfluenceBuilder, instead of using the self.env.app.builder from the assets.py module.

I agree that the plantuml issue does not need to be addressed here and I hope it does not stop this PR  :-)

**Comment by jessetan** [comment]
@jjcaballero this works for my usecase. Thanks for your effort.

File: test/unit-tests/common/test_image_star.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+"""
+
+from sphinxcontrib.confluencebuilder.state import ConfluenceState
+from sphinxcontrib_confluencebuilder_util import ConfluenceTestUtil as _
+from sphinxcontrib.confluencebuilder.builder import ConfluenceBuilder
+import os
+import re
+import unittest
+import tempfile
+import shutil
```
Comment (jdknight) [inline test/unit-tests/common/test_image_star.py]: (optional) Re-ordering these imports into lexicographical order would be a nicety.

File: test/unit-tests/common/test_image_star.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+        with the tested extension. Then the document is built.
+        The expected file is created for every image type just replacing
+        a pattern (@REPLACE_EXTENSION@) in one input file
+        '''
+        dataset = os.path.join(self.test_dir, 'dataset-image-star')
+        image_basename = os.path.basename(self.image_file).split('.')[0]
+        for image_type in ConfluenceBuilder.supported_image_types:
+            specifig_dataset = tempfile.mkdtemp()
+            image_extension = re.sub(r'.*/', '', image_type)
+            os.symlink(os.path.join(dataset, 'index.rst'),
```
Comment (jdknight) [inline test/unit-tests/common/test_image_star.py]: Curious -- will `os.symlink` work in Windows (without additional system platform modifications)?

File: test/unit-tests/common/test_image_star.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        a temporary sourcesdir is created with the original index.rst
+        and a link to the original image file. The link name will end
+        with the tested extension. Then the document is built.
+        The expected file is created for every image type just replacing
+        a pattern (@REPLACE_EXTENSION@) in one input file
+        '''
+        dataset = os.path.join(self.test_dir, 'dataset-image-star')
+        image_basename = os.path.basename(self.image_file).split('.')[0]
+        for image_type in ConfluenceBuilder.supported_image_types:
+            specifig_dataset = tempfile.mkdtemp()
```
Comment (jdknight) [inline test/unit-tests/common/test_image_star.py]: `specific_dataset`

File: test/unit-tests/common/test_image_star.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+    def write_expected_file(self, pattern_file, outfile, actual_extension):
+        ''' writes the content of pattern_file into outfile replacing
+        the image pattern by the actual extension
+        '''
+        with open(pattern_file) as inp, open(outfile, 'w') as out:
+            content = ''.join(inp.readlines())
+            out.write(content.replace(self.extension_pattern, actual_extension))
+
+    def test_all_supported_extensions(self):
```
Comment (jdknight) [inline test/unit-tests/common/test_image_star.py]: A suggestion, if I may, for this test. I looks like you are following the convention found in other tests, were you have a data set, generate results and check on an pre-built expected file for verification; however, you are modifying for each image check the expected file and generating new documentation outputs for each type defined in `supported_image_types`. Why not instead just drop `expected-image-star` completely and generate the entire "expected" document and dummy image types in another temporary directory? This could allow you to have just a single dynamic "expected" document which you only need to perform a single Sphinx build event on (rather than building `len(StandaloneHTMLBuilder.supported_image_types)` times).

File: sphinxcontrib/confluencebuilder/builder.py [hunk 5]
```diff
@@ -22,6 +22,7 @@
 from sphinx.errors import ExtensionError
 from sphinx.util import status_iterator
 from sphinx.util.osutil import ensuredir, SEP
+from sphinx.builders.html import StandaloneHTMLBuilder
```
Comment (jdknight) [inline sphinxcontrib/confluencebuilder/builder.py]: (optional) Re-ordering these imports into lexicographical order would be a nicety.

File: test/unit-tests/common/test_image_star.py (hunk 2 above)
Comment (jjcaballero) [reply test/unit-tests/common/test_image_star.py]: Good point. I didn't know tests must be cross-platform. I learned os.symlink does not work in Windows for all python versions. Eventually, I am copying the file with shutil module and ensuring the created directory is deleted at the end of the test regardless of the result.

File: test/unit-tests/common/test_image_star.py (hunk 3 above)
Comment (jjcaballero) [reply test/unit-tests/common/test_image_star.py]: Addressed.

File: test/unit-tests/common/test_image_star.py (hunk 4 above)
Comment (jjcaballero) [reply test/unit-tests/common/test_image_star.py]: Eventually, I realized that to test an image type, you need an actual image with that image type. It seems the Builder discovers the image candidates according to the actual file type, not only the extension. For every candidate, there is a path/uri assigned. I was testing always the same candidate type (png) with different path assigned.

File: sphinxcontrib/confluencebuilder/builder.py (hunk 5 above)
Comment (jjcaballero) [reply sphinxcontrib/confluencebuilder/builder.py]: Addressed comment. I was not sure if you wanted me to re-order all the imports, I did it, I hope this is ok. I used isort module.

Comment (jdknight) [reply sphinxcontrib/confluencebuilder/builder.py]: Interesting. I was not aware that PEP8 had a recommendation for import ordering. Personally, I am not a fan of this recommendation. I believe (aside from `__future__` requirements) that sorting all imports lexicographically is a much cleaner way to organize imports (for readability); and I never understood why PEP promoted all these extra blank lines.

That being said, maybe I should concede my own personal preferences here. I suspect the majority of Python developers prefer it this way. Also, since Sphinx itself follows this style, I assume it would be good for this extension to follow suit.

File: test/unit-tests/common/test_image_star.py [hunk 6]
```diff
@@ -4,71 +4,88 @@
...
     def setUpClass(cls):
-        cls.extension_pattern = '@REPLACE_EXTENSION@'
-        cls.test_dir = os.path.dirname(os.path.realpath(__file__))
-        cls.image_file = os.path.join(cls.test_dir, 'assets', 'image01.png')
+        test_dir = os.path.dirname(os.path.realpath(__file__))
+        cls.image_file = os.path.join(test_dir, 'assets', 'image01.png')
+        cls.image_basename = cls.file_basename(cls.image_file)
+        cls.datasetdir = tempfile.mkdtemp()
+
+    def tearDown(self):
```
[the remark this answers is not in the document]
Comment (jjcaballero) [reply test/unit-tests/common/test_image_star.py]: Correct! `tearDown` works too because it is called after every single test. But, as datasetdir is created in `setUpClass`, it makes more sense to clean it in `tearDownClass`.

**Comment by jdknight** [comment]
Go time; thanks all! Merged into `master`; available next stable.
